doc/designs/ech-api.md

```diff
@@ -0,0 +1,593 @@
+the ``retry-config`` that may enable a client to use ECH in a subsequent
+connection.
+
+The content of files referred to above must also match the format defined in
+[draft-farrell-tls-pemesni](https://datatracker.ietf.org/doc/draft-farrell-tls-pemesni/).
+
+There are also functions to allow a server to see how many keys are currently
+loaded, and one to flush keys that are older than ``age`` seconds.
+
+```c
+int SSL_CTX_ech_server_get_key_status(SSL_CTX *ctx, int *numkeys);
+int SSL_CTX_ech_server_flush_keys(SSL_CTX *ctx, unsigned int age);
+```
+
+### Split-mode handling
+
+ECH split-mode involves a front-end server that only does ECH decryption and
+then passes on the decrypted inner CH to a back-end TLS server that negotiates
+the actual TLS session with the client, based on the inner CH content. The
+function to support this simply takes the outer CH, indicates whether
+decryption has succeeded or not, and if it has, returns the inner CH and SNI
+values (allowing routing to the correct back-end). Both the supplied (outer)
+CH and returned (inner) CH here include the record layer header.
+
+```c
+int SSL_CTX_ech_raw_decrypt(SSL_CTX *ctx,
+                            int *decrypted_ok,
+                            char **inner_sni, char **outer_sni,
+                            unsigned char *outer_ch, size_t outer_len,
+                            unsigned char *inner_ch, size_t *inner_len,
+                            unsigned char **hrrtok, size_t *toklen);
+```
+
+The caller allocates the ``inner_ch`` buffer, on input ``inner_len`` should
+contain the size of the ``inner_ch`` buffer, on output the size of the actual
+inner CH. Note that, when ECH decryption succeeds, the inner CH will always be
+smaller than the outer CH.
+
+If there is no ECH present in the outer CH then this will return 1 (i.e., the
+call will succeed) but ``decrypted_ok`` will be zero. The same will result if a
+GREASEd ECH is present or decryption fails for some other (indistinguishable)
+reason.
+
+If the caller wishes to support HelloRetryRequest (HRR), then it must supply
+the same ``hrrtok`` and ``toklen`` pointers to both calls to
+``SSL_CTX_ech_raw_decrypt()`` (for the initial and second ClientHello
+messages). When done, the caller must free the ``hrrtok`` using
+``OPENSSL_free()``.  If the caller doesn't need to support HRR, then it can
+supply NULL values for these parameters. The value of the token is the client's
+ephemeral public value, which is not sensitive having being sent in clear in
+the first ClientHello.  This value is missing from the second ClientHello but
+is needed for ECH decryption.
+
+Note that ``SSL_CTX_ech_raw_decrypt()`` only takes a ClientHello as input. If
+the flight containing the ClientHello contains other messages (e.g. a
+ChangeCipherSuite or Early data), then the caller is responsible for
+disentangling those, and for assembling a new flight containing the inner
+ClientHello.
```

Comment:
I think we should provide some more convenient API, i.e. a special TLS_method() that would do all this for the application. But I assume it could be something for future implementation.

Reply:
Yes to that being for future study:-)

Implementing ECH split-mode is pretty odd tbh - it took quite a while to find the right place in the haproxy and nginx code-bases where that could be done. In both cases, the existing application code was essentially acting as a TCP-proxy or similar and wasn't using any TLS APIs so the above seemed to me to make sense. Happy to provide more detail now or later.